Made necessary changes to compile on aarch64

sethkush opened the pull request:
I've made a bit of progress on issue #611 
I've used sse2neon.h from https://github.com/DLTcollab/sse2neon (it's MIT licensed; I hope that's compatible) to get contour to compile on aarch64.
I also made the necessary changes to src/crispy/CMakeLists.txt as well, but I only replaced -maes with the relevant ARM options, so it will break the build on x86_64. Perhaps someone who knows cmake better than I do knows how to switch options based on target architecture?

uspasojevic96 commented:
Thank you, these are the possible architecture checks:

```c
#if defined(__x86_64__) || defined(_M_X64)
        return "x86_64";
        #elif defined(i386) || defined(__i386__) || defined(__i386) || defined(_M_IX86)
        return "x86_32";
        #elif defined(__ARM_ARCH_2__)
        return "ARM2";
        #elif defined(__ARM_ARCH_3__) || defined(__ARM_ARCH_3M__)
        return "ARM3";
        #elif defined(__ARM_ARCH_4T__) || defined(__TARGET_ARM_4T)
        return "ARM4T";
        #elif defined(__ARM_ARCH_5_) || defined(__ARM_ARCH_5E_)
        return "ARM5"
        #elif defined(__ARM_ARCH_6T2_) || defined(__ARM_ARCH_6T2_)
        return "ARM6T2";
        #elif defined(__ARM_ARCH_6__) || defined(__ARM_ARCH_6J__) || defined(__ARM_ARCH_6K__) || defined(__ARM_ARCH_6Z__) || defined(__ARM_ARCH_6ZK__)
        return "ARM6";
        #elif defined(__ARM_ARCH_7__) || defined(__ARM_ARCH_7A__) || defined(__ARM_ARCH_7R__) || defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7S__)
        return "ARM7";
        #elif defined(__ARM_ARCH_7A__) || defined(__ARM_ARCH_7R__) || defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7S__)
        return "ARM7A";
        #elif defined(__ARM_ARCH_7R__) || defined(__ARM_ARCH_7M__) || defined(__ARM_ARCH_7S__)
        return "ARM7R";
        #elif defined(__ARM_ARCH_7M__)
        return "ARM7M";
        #elif defined(__ARM_ARCH_7S__)
        return "ARM7S";
        #elif defined(__aarch64__) || defined(_M_ARM64)
        return "ARM64";
        #elif defined(mips) || defined(__mips__) || defined(__mips)
        return "MIPS";
        #elif defined(__sh__)
        return "SUPERH";
        #elif defined(__powerpc) || defined(__powerpc__) || defined(__powerpc64__) || defined(__POWERPC__) || defined(__ppc__) || defined(__PPC__) || defined(_ARCH_PPC)
        return "POWERPC";
        #elif defined(__PPC64__) || defined(__ppc64__) || defined(_ARCH_PPC64)
        return "POWERPC64";
        #elif defined(__sparc__) || defined(__sparc)
        return "SPARC";
        #elif defined(__m68k__)
        return "M68K";
        #else
        return "UNKNOWN";
        #endif
```

christianparpart commented on src/crispy/CMakeLists.txt:
```diff
@@ -41,7 +41,7 @@ else()
 endif()
 
-target_compile_options(crispy-core PUBLIC -maes)
+target_compile_options(crispy-core PUBLIC -march=armv8-a+fp+simd+crypto+crc)
```
The initial solution (just adding `-maes`) was wrong to begin with. On Windows/MSVC for example this always creates compiler warnings. I am not sure about this change yet (while needed), maybe put it between `if()` conditions? we can also conditionally enable iff the target compiler supports that.

christianparpart replied:
EDIT: Now I've been reading your question. I may come up with a suggestion ASAP (currently busy), maybe someone else's faster otherwise? :)

christianparpart replied:
Looks like https://cmake.org/cmake/help/latest/variable/CMAKE_HOST_SYSTEM_PROCESSOR.html can be used, such as:

```cmake
if(CMAKE_SYSTEM_PROCESSOR STREQUAL x86_64 OR CMAKE_SYSTEM_PROCESSOR STREQUAL amd64)
    target_compile_options(crispy-core PUBLIC -maes)
elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL aarch64) # ARM64
    target_compile_options(crispy-core PUBLIC -march=armv8-a+fp+simd+crypto+crc)
else()
    message(FATAL_ERROR "Target architecture not detected.")
endif()
```

Actual usecase I found on the net: https://github.com/Samsung/netcoredbg/blob/master/detectplatform.cmake#L22-L38

christianparpart commented on src/crispy/StrongLRUCache.h:
```diff
@@ -22,7 +22,11 @@
+#ifdef __x86_64__
 #include <immintrin.h>
+#elif __aarch64__
+#include "sse2neon.h"
```
```suggestion
#include <crispy/sse2neon.h>
```
Maybe?

sethkush replied:
That's what I originally tried, and it worked too. Not sure what's better.

christianparpart reviewed (commented):
Oh wow, happy Sunday morning. Man special thanks @sethkush for that great work. I really very much appreciate that.

Maybe we should mention somewhere where sse2neon.h is coming from, such that an up-to-date version can be easily fetched from?

@sethkush so with this change, can you confirm that it at least sufficiently runs on AArch64 / ARMv8's? :)

christianparpart commented on src/crispy/CMakeLists.txt:
```diff
@@ -41,7 +41,7 @@ else()
 endif()
 
-target_compile_options(crispy-core PUBLIC -maes)
+target_compile_options(crispy-core PUBLIC -march=armv8-a+fp+simd+crypto+crc)
```
```suggestion
if(CMAKE_SYSTEM_PROCESSOR STREQUAL x86_64 OR CMAKE_SYSTEM_PROCESSOR STREQUAL amd64)
    target_compile_options(crispy-core PUBLIC -maes)
elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL aarch64) # ARM64
    target_compile_options(crispy-core PUBLIC -march=armv8-a+fp+simd+crypto+crc)
else()
    message(FATAL_ERROR "Target architecture not detected.")
endif()
```

christianparpart commented on src/crispy/StrongHash.h:
```diff
@@ -21,11 +21,27 @@
 #if defined(__AES__)
     #include <wmmintrin.h>
 #endif
+#elif __aarch64__
```
```suggestion
#elif defined(__aarch64__)
```
Shouldn't we test for "defined" rather then "evaluated to non-zero"?

sethkush replied:
I copied that from here: https://github.com/f1ed/emp/blob/master/emp-tool/utils/block.h
You may be correct though.

christianparpart replied:
I this in the code tree overall you'll find `#if defined(...)` and `#elif defined(...)` the most. So keep it for now (Sadly, C++20 isn't even real-life applicable yet either. :) )

christianparpart commented on src/crispy/StrongHash.h:
```diff
@@ -21,11 +21,27 @@
     #include <wmmintrin.h>
 #endif
+#elif __aarch64__
+#include "sse2neon.h"
```
```suggestion
#include <crispy/sse2neon.h>
```

For the function implementations below, probably mark them `noexcept` and also mention were we've got them from. So that in case of upstream-improvements, it's easier to bugfix them.
I think the original code comes from here? https://github.com/JishinMaster/simd_utils/blob/master/sse2neon_wrapper.h

sethkush replied:
I copied from here: https://github.com/f1ed/emp/blob/master/emp-tool/utils/block.h

christianparpart commented on src/crispy/StrongLRUHashtable.h:
```diff
@@ -22,7 +22,12 @@
 
+#ifdef __x86_64__
 #include <immintrin.h>
+#elif __aarch64__
```
```suggestion
#elif defined(__aarch64__)
```

christianparpart commented on src/crispy/StrongLRUHashtable.h:
```diff
@@ -22,7 +22,12 @@
+#ifdef __x86_64__
 #include <immintrin.h>
+#elif __aarch64__
+#include "sse2neon.h"
```
```suggestion
#include <crispy/sse2neon.h>
```

christianparpart commented on src/crispy/sse2neon.h:
```diff
@@ -0,0 +1,8801 @@
+#ifndef SSE2NEON_H
+#define SSE2NEON_H
+
```
```suggestion
// This file was taken from: https://github.com/JishinMaster/simd_utils/blob/master/sse2neon.h

```
So keeping this file up-to-date with regards to bugfixes but also giving credit, is easier :)

christianparpart commented on src/crispy/sse2neon.h:
```diff
@@ -0,0 +1,8801 @@
+#pragma GCC push_options
+#endif
+#else
+#error "Unsupported target. Must be either ARMv7-A+NEON or ARMv8-A."
```
This is an interesting note. We could add that as a pre-check to the CMake files, too (so the user knows about unsupported platform as early as possible) but also add ARMv7-A+NEON and ARMv8-A to the list of supported target architectures in the project README. :)

sethkush replied:
Also note that the inline functions borrowed from https://github.com/f1ed/emp/blob/master/emp-tool/utils/block.h require the ARMv8 Cryptography Extensions. So I think ARMv7 is out.

christianparpart replied:
> require the ARMv8 Cryptography Extensions. So I think ARMv7 is out.

after googling around, yes, I think ARMv7 is out. Thanks :)

sethkush commented:
> @sethkush so with this change, can you confirm that it at least sufficiently runs on AArch64 / ARMv8's? :)

It is indeed working well on my Pinebook Pro. Transparency seems to be broken, but I can't get gnome-terminal to be transparent either, so I think that's something to do with the GPU.

sethkush commented:
Thanks for all the feedback :smile: 
I'll start working on another iteration cleaning stuff up.

sethkush commented:
Ok, I cleaned up the code and formatted it to fit the project better.
I also incorporated your cmake architecture detection.
I tested the build on both ARMv8 and x86_64, both built and ran.

christianparpart commented:
> Transparency seems to be broken, but I can't get gnome-terminal to be transparent either, so I think that's something to do with the GPU.

background transparency should generally work. The blurred transparrency however requires either Windows 10 or KDE Window Manager (kwin). I'd love to support others at some point in the future (GNOME desktop?).

If background transparency is generally broken, I think that should/could be addressed seperately.

@sethkush, If you think your PR is is mergable, remove the WIP/draft status, so I know that there isn't anything else major coming and we can merge it when ready. :)

christianparpart commented:
@sethkush p.s.: I've just merged another PR to master. Please don't do anything like "git merge master" to merge master into this branch. Either leave it as is, or simply do: "`git fetch && git rebase origin/master`" to rebase your contribution on top of latest master. :)

christianparpart reviewed (approved):
Many thanks @sethkush for your contribution. I really very much appreciate that.
